### ISSUE TITLE
Question about the paper

### ISSUE THREAD

**Issue by nlml** [mpatacchiola/self-supervised-relational-reasoning#2]
Hi there

Thanks for this nice repo + paper.

I just wanted to ask about a couple of curiosities I had re your results, because I wanted to see if this method indeed shows as much promise in outperforming SimCLR as the paper suggests.

I was looking at the supplementary material in the paper. In section B.1 you report linear evaluation accuracy on CIFAR-10 of 61.03%, compared to 60.43% of SimCLR. But if we look at table B.5 in the [SimCLR paper](https://arxiv.org/pdf/2002.05709.pdf), they report CIFAR10 linear evaluation accuracy of 97.7% - much, much higher. Similarly for CIFAR100, the SimCLR authors report 85.9% accuracy, while your implementation of SimCLR is reported achieving just 42.13%

These are pretty big discrepancies, what do you think is the cause? We know SimCLR performs worse at lower batch sizes, which I think you were using, so that probably explains it partially. Also your architecture is smaller. Is that all this is? Or am I missing something? Good chance I am...

Not trying to be mean or anything, just trying to understand your paper well :)

**Comment by nlml** [comment]
Looking again now, I realise table B.5 of SimCLR is about transfer learning from models pretrained on Imagenet -> linear eval on CIFAR10, so this is the wrong comparison. Sorry about that.

However, if we look at Figures B.7 and B.8 in the SimCLR paper, we can see the CIFAR10 pretrained-then-linear-eval'd results of the SimCLR authors that are relevant here. They achieve accuracies of up to 94%. Even with their worst hyperparameter settings they achieve accuracies of 83%+.

This is with a ResNet-50. Your ResNet-56 model (which I presume is larger?) trained using SimCLR achieves 78.75% (Table 7). Does this mean there is something wrong with your re-implement of SimCLR? Because even at low batch sizes the SimCLR authors report much better results.

**Comment by mpatacchiola** [comment]
Hi @nlml thank you for taking the time to read our paper. I have written an answer on Reddit already but I would like to copy it here and add more details to keep this information inside the repository.

Regarding the comparison with SimCLR there are a few major differences between their implementation and ours. In the SimCLR paper they are using: (i) a much larger backbone (ResNet-50 with four hyperblocks), (ii) finetuning instead of linear evaluation, (iii) larger mini-batch size, and (iv) more epochs. We have been unable to replicate their configuration because is very expensive from the computational point of view. They have been using 128 core TPUs v3 (roughly  equivalent to 100/200 GPUs) which is an hardware not many have access to.

In our paper we have followed the idea of flattening out most of the differences (same backbone for all models, same epochs, same evaluation procedure) to have a fair comparison across all methods. Our implementation is open source, so if somebody has access to a larger hardware could try to run the experiments with such a high-level configuration.

**Additional details:**

Regarding the backbone. ResNet-50 is not larger than ResNet-56. We have used the ResNet model with three hyperblocks described in the original ResNet paper, whereas in SimCLR they have adapted the larger model (ResNet-50 with four hyperblock) to run on datasets like CIFAR. Note that, the difference in the number of parameters between the two is significant.
Once again the problem is that ResNet-50 is much larger and we were unable to replicate those results.

The largest backbone we have used is ResNet-34 (with four hyperblocks) in the STL-10 experiments (also in that case our method is superior). Those are reported in the main table in our paper and in the additional material.

**Comment by nlml** [comment]
Thanks for the detailed response. 

In your Table 7 though, you trained on CIFAR10 with a ResNet 56 though right? Is this larger or smaller than ResNet 50 from SimCLR? If it’s larger, why do you think your results are worse than in the SimCLR paper (Figure B7) even when they trained only on CIFAR10 and with a small batch size? Is their ResNet50 there this larger version with more hyperblocks?

**Comment by mpatacchiola** [comment]
Exactly, ResNet-50 (four hyperblocks) is much larger than ResNet-56 (three hyperblocks). We are talking about one order of magnitude more parameters. Here are the figures: ResNet-50 has roughly `24 x 10^6` parameters whereas ResNet-56 with three hyperblocks has `0.8 x 10^6` parameters.

**Comment by nlml** [comment]
Ah ok, that makes sense then. Thanks!
